### DISCUSSION TITLE
Disk space getting eaten up by build cache

### DISCUSSION THREAD

**Discussion by thepuzzlemaster** [dokku/dokku#6726] [Q&A] [answered]
We've got a server which has a front-end nextjs application. A back-end flask application, and a postgres db instance all running on a dokku instance. We recently upgraded (3-4 months or so) from an older version of dokku (I think it was 0.25.6) to the latest. At the same time, we upgraded our hosting server on DigitalOcean to a more recent version of Ubuntu to support the latest version of Dokku.

We've started running into an issue where we run out of disk space when pushing our apps to our dokku box. We don't recall ever running into this on our older version of dokku. We did have a larger disk drive on the older Ubuntu box (250gb vs 80gb now). But we were running on the old box for almost 3 years without ever running into this issue. And we now have run into this disk space issue after only a few months.

We narrowed the issue down to build cache. When running `df` we can see our drive usage tick up after just about every push to dokku.

And running `docker system prune -a` clears up space and gets us from 90%+ back down to 14% usage.

Is this normal? It's unclear if it's related to the updated or just coincidental. But it doesn't seem like something that we should have to monitor as closely as we are having to do now. We didn't see any other issues/discussions around this when searching, but it doesn't seem like anything we're doing is different enough that no one else would be running into this too?

Any thoughts?
Thanks!

**Comment by kminek** [comment]
i can confirm the same, the space gets eaten up on each app deploy. i think dokku should be handling this internally or explictly state in the docs to do system prune from time to time

**Comment by NeoKactus** [comment]
Similar issue here.

I see my disk usage from the last two years go up and down, suggesting that Dokku was cleaning up after itself. We upgraded Dokku last December, and the disk usage only goes up unless we use the docker prune commands.

This is mostly an issue with our development Dokku (v0.33.9) that hosts our review apps. Those apps are constantly being created and destroyed, but there are a bunch of lingering files in `/var/lib/docker/overlay2`.

**Comment by josegonzalez** [comment]
I wonder if we're getting a bunch of _little_ cache hits because of the new dockerfile image builds that are used to tag images now (and add code etc.). I might be able to get away with adding a `--no-cache` flag to the herokuish builder to fix that. Not sure if that's wise to do for the dockerfile builder, but here seems reasonable.

**Comment by bradbajuz** [comment]
I occasionally run `docker buildx prune --all` to free up space by deleting the build cache, which fills up after many deploys.

**Comment by thepuzzlemaster** [comment]
FWIW, at least in our case it does look like the vast majority of space usage is explicitly build cache. We ran into an issue running `docker system prune -a` once. We just ran `docker builder prune` and our storage usage went from 94% to 16%, so we will be using that command moving forward to clear out space, which should still be effective at clearing out this space, but generally safer to run on a running production server.

**Comment by bakatz** [comment]
Just wanted to chime on this one - I hit this very recently and it was initially hard to diagnose. Got an error like "command not found" but the actual issue was no disk space remaining. 

I don't recall exactly which command I ran to clear up space - it was 2 AM and I wanted to go back to sleep. Think it was `docker image prune -a` which cleared 1GB or more of space which then allowed things to go back to normal.

As a frustrating note, when I tried to deploy with no disk space remaining, there was some weird behavior where Dokku deleted my app's config settings entirely. I'm lucky I had them saved in a .env file locally so I was able to quickly reapply them on the next deploy, but it seems strange. Guessing no disk space leads to Dokku thinking the app no longer exists and deleting stuff, I remember seeing some error message to that effect in CI.

I think Dokku should do more to clean up all old docker things (images, build cache, etc). Maybe just include a default cron job that cleans this up once per day, or after every deploy. Would save a lot of pain.

NOTE: I am using the Dockerfile deploy method (git:from-image through the github actions plugin)

**Reply by thepuzzlemaster** [reply]
Oh interesting. @bakatz, it sounds like you hit something very similar to my other post here! https://github.com/dokku/dokku/discussions/6882

**Reply by bakatz** [reply]
> Oh interesting. @bakatz, it sounds like you hit something very similar to my other post here! #6882

Yes, sounds very similar. I tried to look for the log that I saw but it seems Github Actions pruned it. There was definitely a log with something to the effect of "web no longer exists, removing app" which then caused my config to be wiped

**Reply by josegonzalez** [reply]
Dokku never deletes apps unless you run `dokku apps:destroy $APP_NAME`. If you use our official github action, it would only ever delete the review application - which defaults to having the app name and the word "review" in it as shown [here](https://github.com/dokku/ci-docker-image/blob/main/bin/dokku-deploy#L51).

Dokku runs very few processes in the background:

- A process to "retire" (delete) old images and containers that are no longer in use. Data mounted by the containers via the storage plugin is still left on the host where expected, we don't delete those.
- The (I believe optional) `dokku-event-listener` daemon, that listens to container events and rebuilds apps when they hit the max restart limit.

Dokku otherwise does not delete things on it's own.

**Comment by josegonzalez** [comment, marked as answer]
As of Dokku 0.35.0, we now run `docker builder prune` once a day at 4:05am server time. This removes all build cache - docker doesn't expose a way to do this on a per-image basis - which may impact dockerfile builds.

**Reply by thepuzzlemaster** [reply]
Thanks for this. I can confirm that after upgrading to ^0.35.0 our disk space resets nightly and we no longer encounter this issue.
